model_checking/src/lib.rs

```diff
@@ -25,18 +25,46 @@ pub fn mret() {
     ctx.mode = Mode::M;
     ctx.pc = any!();
 
-    let mut sail_ctx = adapters::miralis_to_sail(&ctx);
-    sail_model::execute_MRET(&mut sail_ctx);
+    let sail_ctx = adapters::miralis_to_sail(&ctx);
 
     // Initialize Miralis's own context
     let hw = unsafe { Arch::detect_hardware() };
-    let mut mctx = MiralisContext::new(hw, Plat::get_miralis_start(), 0x1000);
+    let mctx = MiralisContext::new(hw, Plat::get_miralis_start(), 0x1000);
+
+    (ctx, mctx, sail_ctx)
+}
+
+#[cfg_attr(kani, kani::proof)]
+#[cfg_attr(test, test)]
+pub fn mret() {
+    let (mut ctx, mut mctx, mut sail_ctx) = get_miralis_symbolic_context();
 
     ctx.emulate_mret(&mut mctx);
 
+    execute_MRET(&mut sail_ctx);
+
     assert_eq!(
         ctx,
         adapters::sail_to_miralis(sail_ctx),
         "mret instruction emulation is not correct"
     );
 }
+
+#[cfg_attr(kani, kani::proof)]
+#[cfg_attr(test, test)]
+pub fn wfi() {
+    let (mut ctx, mut mctx, mut sail_ctx) = get_miralis_symbolic_context();
+
+    ctx.emulate_wfi(&mut mctx);
+
+    execute_WFI(&mut sail_ctx);
+
+    // This field is used only in Miralis. We set it to false otherwise the assertions fails.
+    ctx.is_wfi = false;
```

Comment:
Note: we could also verify that the injection of interrupts is working as expected. There is a `dispatchInterrupt` function in Sail, maybe we could use it somehow for verification as we have a similar function.